src/tls.rs

```diff
@@ -352,14 +384,40 @@ impl Accept for TlsAcceptor {
         cx: &mut Context<'_>,
     ) -> Poll<Option<Result<Self::Conn, Self::Error>>> {
         let pin = self.get_mut();
-        match ready!(Pin::new(&mut pin.incoming).poll_accept(cx)) {
-            Some(Ok(sock)) => {
-                let session = ServerSession::new(&pin.config.clone());
-                // let tls = Arc::new($this.config);
-                return Poll::Ready(Some(Ok(TlsStream::new(sock, session))));
+        // fast path for not upgrading
+        if !pin.upgrade {
```

Comment:
There is some duplicate code here to keep the default path as minimal as it was before.
I could refactor the duplicate code out into a function if thats better?

Reply:
I played around with hit a little, it's done.